Fix backpropagation of conditionals when the type is later widened

Keno opened the pull request:
Consider the following example:
```
f() = (1, 1)
f(s) = i == 1 ? (1, 2) : nothing
function foo()
    next = f()
    while next !== nothing
        next = f(Core.getfield(next, 2))
    end
end
```

This was causing unnecessary allocations, because inference failed
to inform codegen that the reference to `next` inside the loop was
guaranteed not to be `Nothing`. Ironically if `f()` could also have
returned nothing, inference was able to figure this out, thanks to
pass the result of the comparison was inferred as `Const(true)`,
and on the second pass as `Conditional(next, Tuple{Int, Int},
Const(nothing))`, the tmerge of which is `Bool`. Fix that by
implementing Jameson's suggestion in #25261, and instead infering
the first pass as `Conditional(next, Tuple{Int, Int}, Union{})`.
To make this work we also adjust the lattice relation to have
`Conditional(next, T, Union{}) ⊑ Const(true)`. The result is that
the result of the comparison does not get widened to `Bool` as
quickly, retaining the type information on `next` and allowing
codegen to emit efficient code.

JeffBezanson commented on base/inference.jl:
```diff
@@ -2940,6 +2948,8 @@ function ⊑(@nospecialize(a), @nospecialize(b))
         if isa(b, Conditional)
             return issubconditional(a, b)
+        elseif isa(b, Const) && isa(b.val, Bool)
+            return maybe_extract_const_bool(a) === b.val
```
Should add corresponding code for `Const ⊑ Conditional`.

It also seems we don't have `Bool ⊑ Conditional`, but shouldn't that be true (unless one branch is `Bottom`)?

Keno replied:
Frankly, I'm not sure. `Conditional` is a bit of an odd element here. Are you suggesting that:
```
for (a, b) in ((Conditional(_, T, S), Bool), (Conditional(_, T, Union{}), Const(true)))
@test a ⊑ b
@test b ⊑ a
end
```
should hold?

JeffBezanson replied:
After drawing out the lattice, I see you might be right. It seems `Conditional(_, T, Union{})` is "slightly smaller" than `Const(true)`.

It's unintuitive that `Const(true) ⊑ Conditional(_, T, T)` doesn't hold, which is probably what's confusing me.

Keno replied:
> It's unintuitive that Const(true) ⊑ Conditional(_, T, T) doesn't hold, which is probably what's confusing me.

You're right, if both branches are the same I would expect it to be equivalent to Bool in essentially every way, because that basically says, independent of the condition, that `var` is constrained by `T`, so it really should just have been done that way in the first place.

vtjnash replied:
Yes, currently, we have that `Conditional(_, T, T) === Bool` and `Conditional(_, T, S) ⊑ Bool`

This PR is proposing making `Conditional(_, T, Union{}) ⊑ Const(true)` as a special case (currently the common lattice element of those two is `Bool`, but the former also should currently never appear)

Keno replied:
By `===` in the first one I assume you mean `===(a,b) = a ⊑ b && b ⊑ a`. If so, is that actually true for the current code?

vtjnash replied:
It's only true in theory, since we don't check for it specifically in `⊑`. In practice, I just usually tried to avoid producing that, so that I wouldn't have to worry about it, and it doesn't have any effect to carry around this Conditional.

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/25557#issuecomment-357553463):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/82f67341291f07cffaca73bf79e9852f1aa002b7/f81c623_vs_9924f79/report.md)
cc @ararslan

ararslan commented:
I just got the fix in for the change to `find`, so Nanosoldier should work now (assuming this PR doesn't somehow break BaseBenchmarks).

@nanosoldier `runbenchmarks(ALL, vs=":master")`

ararslan commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25557#issuecomment-357558071) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/4a85ee05c8b5211ca96590438eb15d00dfcd1977/f81c623_vs_9924f79/report.md). cc @ararslan

Keno commented:
Performance regressions seem real, unfortunately.

Keno committed:
Hopefully fix performance problems

There's a couple of places in the optimization passes that benefit
from Const information. However, they don't really need to care
about the new Conditional lattice elements, which only matter to
inference. Simply widen these before we get to optimization to
hopefully recover any lost performance.

Keno committed:
Fix backpropagation of conditionals when the type is later widened

Consider the following example:

```
f() = (1, 1)
f(s) = i == 1 ? (1, 2) : nothing
function foo()
    next = f()
    while next !== nothing
        next = f(Core.getfield(next, 2))
    end
end
```

This was causing unnecessary allocations, because inference failed
to inform codegen that the reference to next inside the loop was
guaranteed not to be Nothing. Ironically if f() could also have
returned nothing, inference was able to figure this out, thanks to
pass the result of the comparison was inferred as Const(true),
and on the second pass as Conditional(next, Tuple{Int, Int}, Const(nothing)),
the tmerge of which is Bool. Fix that by
implementing Jameson's suggestion in #25261, and instead infering
the first pass as Conditional(next, Tuple{Int, Int}, Union{}).
To make this work we also adjust the lattice relation to have
Conditional(next, T, Union{}) ⊑ Const(true). The result is that
the result of the comparison does not get widened to Bool as
quickly, retaining the type information on next and allowing
codegen to emit efficient code.

ararslan commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25557#issuecomment-358022881) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/29cdd76ed1cf5acf0badf7aa5674a910dd5c42ad/5af55df_vs_19429e2/report.md). cc @ararslan

Keno commented:
While I was able to originally reproduce some of these performance regressions, I no longer can, so I 'm gonna recommend merging this once CI has a reasonable level of confidence.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25557#issuecomment-358088189) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/4b40076b8a63245a9234a891fbd77c79d3442cc8/a2ad240_vs_aea9155/report.md). cc @ararslan
